File: presences/src/main/java/fr/openent/presences/service/ActionService.java
package fr.openent.presences.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.List;

public interface ActionService {
    /**
     * get actions
     *
     * @param structureId Structure Identifier
     * @param handler     Function handler returning data
     */
    void get(String structureId, Handler<Either<String, JsonArray>> handler);

    /**
     * create action setting
     *
     * @param actionBody actionBody fetched
     * @param handler    Function handler returning data
     */
    void create(JsonObject actionBody, Handler<Either<String, JsonObject>> handler);

    /**
     * put reason setting
     *
     * @param actionBody actionBody fetched
     * @param handler    Function handler returning data
     */
    void put(JsonObject actionBody, Handler<Either<String, JsonObject>> handler);

    /**
     * delete reason setting
     *
     * @param actionId action identifier
     * @param handler  Function handler returning data
     */
    void delete(Integer actionId, Handler<Either<String, JsonObject>> handler);


    /**
     * get last action
     *
     * @param events  event list
     * @param handler Function handler returning data
     */
    void getLastAbbreviations(List<Integer> events, Handler<Either<String, JsonArray>> handler);
}

File: app-presences/src/main/resources/public/ts/services/ActionService.ts
import {ng} from 'entcore'
import http, {AxiosResponse} from 'axios';
import {Action, ActionRequest} from "../models/Action";

export interface ActionService {
    getActions(structureId: string): Promise<Action[]>;

    // Settings
    create(actionBody: ActionRequest): Promise<AxiosResponse>;

    update(actionBody: ActionRequest): Promise<AxiosResponse>;

    delete(actionId: number): Promise<AxiosResponse>;
}

export const actionService: ActionService = {
    getActions: async (structureId: string): Promise<Action[]> => {
        try {
            const {data} = await http.get(`/presences/actions?structureId=${structureId}`);
            return data;
        } catch (err) {
            throw err;
        }
    },

    // Settings
    create: (actionBody: ActionRequest): Promise<AxiosResponse> => {
        return http.post(`/presences/action`, actionBody);
    },

    update: (actionBody: ActionRequest): Promise<AxiosResponse> => {
        return http.put(`/presences/action`, actionBody);
    },

    delete: (actionId: number): Promise<AxiosResponse> => {
        return http.delete(`/presences/action?id=${actionId}`);
    }
};
export const ActionService = ng.service('ActionService', (): ActionService => actionService);